applications/rag/frontend/container/cloud_sql/cloud_sql.py

```diff
@@ -52,32 +55,79 @@ def getconn() -> pymysql.connections.Connection:
-      query_emb = transformer.encode(query_text).tolist()
-      query_request = "SELECT id, text, text_embedding, 1 - ('[" + ",".join(map(str, query_emb)) + "]' <=> text_embedding) AS cosine_similarity FROM " + TABLE_NAME + " ORDER BY cosine_similarity DESC LIMIT 5;"
-      query_results = conn.execute(sqlalchemy.text(query_request)).fetchall()
-      conn.commit()
-
-      if not query_results:
-        message = f"Table {TABLE_NAME} returned empty result"
-        raise ValueError(message)
-      for row in query_results:
-        print(row)
-    except sqlalchemy.exc.DBAPIError or pg8000.exceptions.DatabaseError as err:
-      message = f"Table {TABLE_NAME} does not exist: {err}"
-      raise sqlalchemy.exc.DataError(message)
-    except sqlalchemy.exc.DatabaseError as err:
-      message = f"Database {INSTANCE_CONNECTION_NAME} does not exist: {err}"
-      raise sqlalchemy.exc.DataError(message)
-    except Exception as err:
-      raise Exception(f"General error: {err}")
-
-  return query_results[0][1]
\ No newline at end of file
+def create_sync_postgres_engine():
+    engine = PostgresEngine.from_instance(
+        project_id=PROJECT_ID,
+        region=REGION, 
+        instance=INSTANCE,
+        database=DB_NAME,
+        user=DB_USER,
+        password=DB_PASS,
+        ip_type=IPTypes.PRIVATE        
+    )
+    engine.init_chat_history_table(table_name=CHAT_HISTORY_TABLE_NAME)
+    return engine
+
+#TODO replace this with the Cloud SQL vector store for langchain,
+#   once the notebook also uses it (and creates the correct schema)
+class CustomVectorStore(VectorStore):
+    @classmethod
+    def from_texts(
+        cls,
+        texts: List[str],
+        embedding: Embeddings,
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ):
+        raise NotImplementedError
+
+    def __init__(self, embedding: Embeddings, engine: Engine):
+        self.embedding = embedding
+        self.engine = engine
+
+    @property
+    def embeddings(self) -> Embeddings:
+        return self.embedding
+
+
+    # TODO implement
+    def add_texts(self, texts: Iterable[str], metadatas: List[dict] | None = None, **kwargs: Any) -> List[str]:
+        raise NotImplementedError
+    
+    #TODO implement similarity search with cosine similarity threshold
+
+    def similarity_search(self, query: dict, k: int = 4, **kwargs: Any) -> List[Document]:
+        print("ENTERING similarity_search")
+        with self.engine.connect() as conn:
+            try:
+                q = query["question"]
+                print(f"ENTERING query embedding for '{q}'")
+                # embed query & fetch matches
+                query_emb = self.embedding.embed_query(q)
+                print(f"GOT embedding of length: {len(query_emb)}")
+                emb_str = ",".join(map(str, query_emb))
+                query_request = f"""SELECT id, text, 1 - ('[{emb_str}]' <=> text_embedding) AS cosine_similarity 
```

Comment:
I think we should get all and rank it then do the cut off, limit k may have a bad performance issue here.

Reply:
so far with the small dataset, I haven't noticed any perf issues. If it comes up w/ a bigger dataset, then I'm all for improving this query.